Fix code style, improve Docker config and restore unbinarize easter egg

ajayat opened the pull request:
- Fix code style violations (from pre-commit hook)
- Minor changes on Docker config
- Re-add unbinarize option, by replacing inserting an invisible space after '@' to avoid mention

ajayat committed:
Fix code style

ajayat committed:
Add safe unbinarize easteregg

ajayat committed:
Update docker config

ajayat committed:
Remove alias for models

Triw-12 commented on mp2i/cogs/suggestions.py:
```diff
@@ -171,48 +179,46 @@ async def suggestions(self, ctx, state: str) -> None:
 
         if state == "accepted":
-            embed = discord.Embed(title=f"Suggestions acceptées", colour=0x77B255, timestamp=datetime.now())
+            embed = discord.Embed(title="Suggestions acceptées", color=0x77B255)
```
Tu as remplacé juste le premier colour en color mais pas les autres

Triw-12 commented on mp2i/cogs/suggestions.py:
```diff
@@ -171,48 +179,46 @@ async def suggestions(self, ctx, state: str) -> None:
-                    inline=False,
-                )
+                guild = GuildWrapper(ctx.guild)
+                msg = await guild.suggestion_channel.fetch_message(suggest.message_id)
```
Il faudrait sûrement mettre un try execpt ou une autre protection pour éviter de faire planter la commande si le message n'est pas trouvé (si il a été supprimé par exemple par un moderateur/administrateur)

ajayat replied:
Tu as tout à a fait raison, j'ai laissé la chose telle qu'elle était (j'ai juste refactorisé) mais c'est effectivement un potentiel bug

Triw-12 replied:
Oui, c'est ma faute, j'y ai pensé que après

Triw-12 commented on mp2i/cogs/suggestions.py:
```diff
@@ -87,33 +88,40 @@ async def close_suggestion(self, payload) -> None:
+        if str(payload.emoji) == accept.emoji:
+            state = "accepted"
+            embed = discord.Embed(
+                colour=0x77B255,
```
Si tu préfère utiliser color à la place de colour, autant le remplacer partout je pense

Triw-12 commented on Dockerfile:
```diff
@@ -4,8 +4,6 @@ FROM python:3.9-slim-buster
 ENV PYTHONUNBUFFERED=1 \
     PIPENV_NOSPIN=1
-# Configure timezone
-ENV TZ Europe/Paris
```
Il ne faudrait pas aussi l'enlever dans le .env du README.md du coup ?

ajayat replied:
Nop, justement je l'enlève d'ici car il est deja dans le .env du README qui est chargé dans le conteneur (information redondante donc)

Triw-12 commented on mp2i/cogs/commands.py:
```diff
@@ -205,5 +205,19 @@ async def leaderboard(self, ctx, rmax: Optional[int] = 10) -> None:
         embed = discord.Embed(colour=0x2BFAFA, title=title, description=content)
         await ctx.send(embed=embed)
 
+    @Cog.listener("on_message")
+    async def unbinarize(self, msg: discord.Message):
+        """
+        Vérifie si le message est un texte binaire et le convertit en ASCII.
+        """
+        if not re.fullmatch(r"([01]{8}\s?)+", msg.content):
+            return
+
+        binary = re.findall("[01]{8}", msg.content)
+        # Convert to ASCII and replace @ by @\u200b to avoid mention
+        text = "".join(chr(int(b, 2)) for b in binary).replace("@", "@\u200b")
+        await msg.reply(text)
+
+
```
Il est préférable d'utiliser les `allowed_mentions`

Références API :
- https://discordpy.readthedocs.io/en/stable/api.html?highlight=send%20message#discord.AllowedMentions
- https://discordpy.readthedocs.io/en/stable/api.html?highlight=send%20message#discord.PartialMessageable.send

On peut aussi changer les allowed_mentions par défaut sur le bot : https://discordpy.readthedocs.io/en/stable/api.html?highlight=send%20message#discord.Client.allowed_mentions

ajayat replied:
Je pense que le mieux est de le faire au niveau du message, on peut avoir besoin des mentions par le bot dans d'autres contextes

Triw-12 commented on mp2i/cogs/commands.py:
```diff
@@ -205,5 +205,19 @@ async def leaderboard(self, ctx, rmax: Optional[int] = 10) -> None:
         embed = discord.Embed(colour=0x2BFAFA, title=title, description=content)
         await ctx.send(embed=embed)
 
+    @Cog.listener("on_message")
+    async def unbinarize(self, msg: discord.Message):
+        """
+        Vérifie si le message est un texte binaire et le convertit en ASCII.
+        """
+        if not re.fullmatch(r"([01]{8}\s?)+", msg.content):
+            return
+
+        binary = re.findall("[01]{8}", msg.content)
+        # Convert to ASCII and replace @ by @\u200b to avoid mention
+        text = "".join(chr(int(b, 2)) for b in binary).replace("@", "@\u200b")
+        await msg.reply(text)
+
+
```
J'ai parlé sur le discord discord.py du problème, et il peut en effet avoir plusieurs autres failles :

- Mettre un cooldown : le spamming peut entraîner la limitation du bot et éventuellement son bannissement temporaire
- Échapper/supprimer le markdown pour éviter les abus potentiels de bugs, etc.
- Ajoutez une limite de caractères, par exemple 1000, car les robots ne peuvent envoyer que 2000 caractères et pour éviter les spams.
- Supprimer ou vérifier les insultes et ignorer -> le bot peut bypass l'automod

ajayat replied:
Normalement tout est corrigé, il y a au moins un filtrage des bad words en anglais, on pourra en trouver / faire un pour le français.
De toute manière, les même sanctions que pour le contournement d'automod s'appliquent

Triw-12 commented on mp2i/cogs/commands.py:
```diff
@@ -205,5 +205,19 @@ async def leaderboard(self, ctx, rmax: Optional[int] = 10) -> None:
         embed = discord.Embed(colour=0x2BFAFA, title=title, description=content)
         await ctx.send(embed=embed)
 
+    @Cog.listener("on_message")
+    async def unbinarize(self, msg: discord.Message):
+        """
+        Vérifie si le message est un texte binaire et le convertit en ASCII.
+        """
+        if not re.fullmatch(r"([01]{8}\s?)+", msg.content):
+            return
+
+        binary = re.findall("[01]{8}", msg.content)
+        # Convert to ASCII and replace @ by @\u200b to avoid mention
+        text = "".join(chr(int(b, 2)) for b in binary).replace("@", "@\u200b")
+        await msg.reply(text)
+
+
```
![image](https://github.com/user-attachments/assets/7b25844a-488e-4659-a7ba-79a38923c1ad)

Ne pas permettre que le bot se réponde à lui-même

Je pense que pour que ce soit plus simple, il faudrait juste mettre un ` au début et à la fin et aussi les escapes, ça régera déjà quelques problèmes que j'ai énoncé.

ajayat replied:
Très drôle j'y avais pas pensé.
Tout simplement on vérifie que le message ne provient pas d'un bot

ajayat commented:
Je rajoute un problème potentiel:
- les messages qui commencent par =, pour empêcher l'exécution de commande par le bot

Triw-12 commented:
> Je rajoute un problème potentiel:
>     -  les messages qui commencent par =, pour empêcher l'exécution de commande par le bot

Alors normalement non : `The bot cannot execute it's own commands, however it does trigger its own on_Message Events`
Mais du coup il peut plutôt faire les commandes pour l'autre bot LaTex, mais c'est pas si grave après

ajayat committed:
Fix color inconsistency and add try/except

ajayat committed:
Disable mentions in unbinarize reply

ajayat committed:
Add import alias:
database -> db
GuildModel -> GM
SuggestionModel -> SM
SanctionModel -> SM
MemberModel -> MM

ajayat commented:
Petit commit de style en même temps: 
Autant database -> db, on ne perd pas en clarté, par contre je suis pas sûr pour GuildModel -> GM etc, t'en pense quoi ? (Etant donné que c'est utilisé souvent, ça rend le code plus concis)

ajayat committed:
Ignore unbinarize from bot users

Triw-12 commented:
> Petit commit de style en même temps: Autant database -> db, on ne perd pas en clarté, par contre je suis pas sûr pour GuildModel -> GM etc, t'en pense quoi ? (Etant donné que c'est utilisé souvent, ça rend le code plus concis)

Je pense que c'est pas si dérangant avec les alias. Après le faire que SM ce soit SuggestionModel et SanctionModel c'est un peu dommage, mais normalement il suffit de regarder le nom du fichier pour savoir. Après si un jour pour une raison quelconque les deux tables sont utilisés dans un même fichier, ça deviendra dérangant

ajayat commented:
> > Petit commit de style en même temps: Autant database -> db, on ne perd pas en clarté, par contre je suis pas sûr pour GuildModel -> GM etc, t'en pense quoi ? (Etant donné que c'est utilisé souvent, ça rend le code plus concis)
> 
> Je pense que c'est pas si dérangant avec les alias. Après le faire que SM ce soit SuggestionModel et SanctionModel c'est un peu dommage, mais normalement il suffit de regarder le nom du fichier pour savoir. Après si un jour pour une raison quelconque les deux tables sont utilisés dans un même fichier, ça deviendra dérangant

Normalement, notre code doit être suffisamment modularisé pour que ça n'arrive jamais! Tu trouves qu'on perd un peu en clarté avec les alias SM, GM etc... ? Si oui je peux les enlever, car j'hésite.

ajayat committed:
Use profanity censoring

Triw-12 commented:
> > > Petit commit de style en même temps: Autant database -> db, on ne perd pas en clarté, par contre je suis pas sûr pour GuildModel -> GM etc, t'en pense quoi ? (Etant donné que c'est utilisé souvent, ça rend le code plus concis)
> > 
> > 
> > Je pense que c'est pas si dérangant avec les alias. Après le faire que SM ce soit SuggestionModel et SanctionModel c'est un peu dommage, mais normalement il suffit de regarder le nom du fichier pour savoir. Après si un jour pour une raison quelconque les deux tables sont utilisés dans un même fichier, ça deviendra dérangant
> 
> Normalement, notre code doit être suffisamment modularisé pour que ça n'arrive jamais! Tu trouves qu'on perd un peu en clarté avec les alias SM, GM etc... ? Si oui je peux les enlever, car j'hésite.

Après je peux pas trop répondre parce que je suis par défaut biaisé parce que je connais bien le projet. Mais je pense que si j'arrivais sur le projet en n'y connaissant pas grand chose, le fait d'utiliser des alias rend quand même les choses compliqué à comprendre facilement.... Personnellement, je trouve pas trop dérangant que les lignes soient un peu plus longue si ça permet que ce soit plus clair (donc de ne pas utiliser des alias)

ajayat commented:
En effet, pour plus de clarté je vais enlever les alias.
Concernant la feature, je pense qu'on est bon, je te laisse tester de ton côté et tu me dis 👍

ajayat committed:
Revert "Add import alias:"

This reverts commit 94a7db73219cd91adc2a9c3bff191c3d64f5875a.

ajayat committed:
Refactor: Update database imports and usage in suggestions.py

ajayat committed:
Add toxicity detection [untested]

ajayat committed:
Refactor: Update Docker configuration and model handling; enhance toxicity detection logging

ajayat committed:
Update Dockerfile

ajayat committed:
Merge remote-tracking branch 'origin/develop' into feat/unbinarize

ajayat committed:
Merge remote-tracking branch 'origin/develop' into feat/unbinarize

ajayat committed:
Update Docker config and README

ajayat committed:
Merge remote-tracking branch 'origin/develop' into feat/unbinarize

Alessevan commented:
As-tu mesuré la consommation en RAM de la version du bot que tu proposes ?

ajayat commented:
> As-tu mesuré la consommation en RAM de la version du bot que tu proposes ?

Excellente remarque, non pas encore, je vais faire les tests demain

Triw-12 commented:
> As-tu mesuré la consommation en RAM de la version du bot que tu proposes ?

![image](https://github.com/user-attachments/assets/038e013a-3f7b-43e9-b5da-71c228da749a)


J'ai fait les tests pour voir, et je me retrouve avec une consummation de environ 1,5 Go en continu. Ce n'est pas tenable sur un VPS de 2Go. Et j'ai regardé du coup pour des optimisations et il n'y a rien qui me convint vraiment : 
- Mettre la LLM uniquement quand on en a besoin risque de poser de gros lags à chaque fois qu'un message est posté
- Même en faisant du swap offloading, les performances seront bien trop basses pour avoir un LLM efficace. Après c'est à tester mais après que je me sois renseigné, ça me semble vraiment très compliqué de mettre un LLM sur un VPS

ajayat commented:
Parfait ! La pratique confirme la théorie :
Le modèle de base sur l'architecture `XLM-roBERTa-base` qui consomme en inférence environ 1.5Go de RAM.
Cependant, actuellement le modèle est chargé en entier (poids, buffets, cache, tenseurs) alors que pour l'inférence seuls les poids suffisent. Je vais voir si je peux réduire la taille du modèle à 1Go environ.
Ce qui le rend utilisable sur le VPS. @Triw-12 @Alessevan

ajayat commented:
Par ailleurs, pour l'automod, je propose cette méthode pour déterminer le seuil de détection :
1. Récupérer un dataset de test ou en créer un (plus personnalisé) comme ça par ex : 
- Envoyer à la communauté un formulaire pour rédiger des phrases insultantes, ambiguës ou normales.    
- Demander aux modérateurs de cocher chaque phrase qui devrait, selon eux, être modérée.    
2. Choisir le seuil de décision optimal :
- Tracer la courbe ROC (Receiver Operating Characteristic).    
- Prendre le seuil correspondant au point le plus proche de (0,1) sur cette courbe (maximise la détection tout en minimisant les fausses alertes).    
=> L’augmenter si l’on souhaite réduire davantage le taux de faux positifs (FPR), ie le nombre de messages modérés à tort

ajayat commented:
Si besoin, on pourrait aussi réfléchir à étendre le VPS de 2Go à 4Go de RAM pour 2.97€/mois.
Au delà du bot, ça pourra service pour de futurs services sur le VPS (comme Gitea)

Triw-12 commented:
> Parfait ! La pratique confirme la théorie : Le modèle de base sur l'architecture `XLM-roBERTa-base` qui consomme en inférence environ 1.5Go de RAM. Cependant, actuellement le modèle est chargé en entier (poids, buffets, cache, tenseurs) alors que pour l'inférence seuls les poids suffisent. Le modèle devrait donc consommer un peu moins de 1Go de Ram avec les poids seuls. Ce qui le rend utilisable sur le VPS. @Triw-12 @Alessevan

![image](https://github.com/user-attachments/assets/5a321707-d4f2-45ba-b889-2f41d10216b6)

Actuellement le VPS utilise 750 Mo de RAM "au repos". Rajouter entre 1Go et 1,5Go de RAM sur le serveur n'est donc pas tennable, ça laisse maximum 200 Mo pour faire toutes les autres actions, ce qui risque de ralentir à la fois les commandes du bot et l'accès au site web.

ajayat commented:
Sinon, on pourrait utiliser une clé d'API pour un modèle comme BERT ou GPT , et là on est tranquille niveau ressource !
Mais on a des quotas, et il faut se décider sur lequel on utilise éthiquement parlant.

ajayat commented:
J'ai fouillé un peu els modèles sur HuggingFace, j'ai trouvé des modèles convenables jusqu'a 3x plus légers que roberta

ajayat committed:
feat: change automod model and refactor schools:
 Add optimization with optimum

ajayat commented:
## Changements importants

### Nouveau modèle et optimisations
J'ai changé le modèle pour celui là sur [HuggingFace](https://huggingface.co/citizenlab/distilbert-base-multilingual-cased-toxicity). Puis je l'ai optimisé avec `optimum` et exporté au format Onnx, puis ajouté avec Git LFS.
Il prends 250 Mo sur le disque et environ **400-500 Mo dans la RAM**. Le chargement du modèle et l'inférence sont instantanées.

### Optimisation des packages Python
J'ai aussi fais en sorte de coder la pipeline à la main pour éviter d'utiliser PyTorch (qui même en version cpu prend 700 Mo). Ainsi l'ensemble des packages Python ne prend que 250 Mo environ, le container Docker fait 500 Mo.

### Précision du modèle
Après les premiers tests, le modèle est moins précis que `roberta` sur le français, mais se débrouille très bien en anglais. Il a un recall assez faible, mais ça évite au moins les faux positifs.

Pour résumer, il est désormais suffisament performant pour tourner sur le VPS et inférer sur tous les messages (en offline).

Alessevan commented:
Avec cette version, j'ai cette erreur au démarrage du bot :
```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/bot/mp2i/__main__.py", line 6, in <module>
    asyncio.run(bot.run())
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/bot/mp2i/bot.py", line 31, in run
    for cog in resolver.find_available_cogs():
  File "/bot/mp2i/utils/resolver.py", line 21, in find_available_cogs
    imported = import_module(module.name)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/bot/mp2i/cogs/commands.py", line 18, in <module>
    from mp2i.utils import youtube, automod
  File "/bot/mp2i/utils/automod.py", line 95, in <module>
    classifier = ToxicityClassifier(
  File "/bot/mp2i/utils/automod.py", line 38, in __init__
    self._tokenizer = self._load_tokenizer()
  File "/bot/mp2i/utils/automod.py", line 46, in _load_tokenizer
    return Tokenizer.from_file(tokenizer_path.as_posix())
Exception: expected value at line 1 column 1
```
Je suis à jour par rapport à la branche distante, pas de modifications locales et je l'ai lancé par Docker.

ajayat commented:
@Alessevan as-tu essayé `git lfs pull` ?

Alessevan commented:
Ok, c'est bon, `git-lfs` n'est pas de base sur ma machine.

ajayat commented:
Mais merci de la remarque, je vais l'ajouter dans le README

ajayat committed:
Update README.md

ajayat committed:
feat: Add automod for message moderation

ajayat commented:
Abandon de la feature `unbinarize`.
Voir https://github.com/prepas-mp2i/mp2i-discord-bot/pull/22 pour les refactorisations et l'automod.
